### PR TITLE
feature/extended-selenium-grid-containers

### PR DESCRIPTION
**Pull request by sskorol**
Revised Selenium containers for better scaling and avoiding JUnit / WebDriver dependencies:
- created GenericGridContainer to allow better flexibility in creating dedicated SeleniumHQ containers;
- created SeleniumHubContainer for standalone usage with dedicated nodes; note that it doesn't support VNC / video recording;
- created SeleniumNodeContainer with VNC / video recording support for standalone usage with SeleniumHubContainer;
- created SeleniumStandaloneContainer as an alternative to BrowserWebDriverContainer, but without WebDriver dependency;
- created 2 enums: Browser / SeleniumImage for better containers management;
- created SeleniumHttpUtils for checking if container is ready to be used by tests;
- added brief overview into webdriver_containers.md.
- added required dependencies and developer's info

### CODE REVIEW

File: modules/selenium/pom.xml [hunk 1]
```diff
@@ -33,5 +33,26 @@
             <artifactId>jetty</artifactId>
             <version>6.1.25</version>
         </dependency>
+
+        <!-- Concise and flexible API for polling hub container, until it's up and running -->
```
Comment (sskorol) [inline modules/selenium/pom.xml]: forgot to change a comment... this lib is used for all containers polling, not only for hub (see overridden `waitUntilContainerStarted`)

**Comment by sskorol** [comment]
Note that I haven't added high level tests for these updates to avoid additional non-related project dependencies, e.g. testng, object pool, webdriver factory, etc. Not quite sure that root project is a good place for this. I believe it's better to add corresponding tests into `examples` project. They're actually ready, just waiting when new containers are merged.

**Comment by sskorol** [comment]
@rnorth any comments / concerns on this?

**Comment by rnorth** [comment]
Hi
Sorry for the slow reply - this seems like a big piece of work, so thank you. I've not yet had a chance to try it out, but should hopefully have a good bit of time over the upcoming weekend. I'll try and give some feedback then.

One thing is that I really would like to find a way to get good test coverage into this as part of the regular build/CI process. I'll try and have a look at your examples though and see what the impact would be. If it's just a case of adding <test> scoped JARs to a module that would be fine, though.

Thanks
Richard

**Comment by sskorol** [comment]
Hi Richard,

I'll push corresponding examples today. And then we could discuss what could be included or not.

**Comment by sskorol** [comment]
@rnorth I'm not quite sure, how trevis / circle CI are configured, but seems like that code, which tries to poll nodes doesn't work on CI. And I have no idea why, to be honest.

Local `clean install` 100% [passing](https://drive.google.com/open?id=0B1ORH0iP6-geVzE0VzMybnlLa28).

Interesting thing that new standalone container's tests work fine on trevis / circle. The problem is in dedicated grid, when we try to ask hub, if nodes are connected with the following code:

```java
	@Override
	protected void waitUntilContainerStarted() {
		given().ignoreExceptions()
				.atMost(POLLING_TIMEOUT, TimeUnit.SECONDS)
				.pollInterval(POLLING_INTERVAL, TimeUnit.SECONDS)
				.until(() -> {
					final boolean isConnected = checkHostState(new HttpHost(getHubAddress(), getHubPort()),
							new BasicHttpRequest("GET", getProxyUrl().toExternalForm()));
					logger().info("Obtained a connection to ({}) = {}", getProxyUrl(), isConnected);
					return isConnected;
				});
	}

	private URL getProxyUrl() {
		return Try.of(() -> new URL("http", getHubAddress(), getHubPort(), "/grid/api/proxy?id=http://" +
				getInternalIpAddress() + ":" + DEFAULT_NODE_PORT)).getOrElseTry(() -> new URL(DEFAULT_PROXY_URL));
	}
```

Where `http://hubIp:port/grid/api/proxy?id=http://nodeIp:port` is the way how we can check active node, if it's ready for interaction. For these tests we're getting 60 sec timeout on connection waiting.

Any thoughts, how trevis / circle CI internal network is configured to be able to poll hub / node status correctly?

Would be greatly appreciated if you could give a valid direction.

Thanks,
Sergey

**Comment by rnorth** [comment]
Hi Sergey,
Re the CI test failures, this will probably take a bit more investigation. Those btrfs warnings are normal for CircleCI, and relate to extra constraints CircleCI has for its docker setup. However, I also get a test setup failure locally when running the TestNG example tests so there may be something more to it ([details below](#s1)).

Is this understanding of your main goals and approach correct?
- Use of a Selenium grid master node and a pool of multiple containers - the goal being to provide enough browser instances to support parallel testing, presumably including having the same test be run on both Chrome and Firefox concurrently.
- Support for TestNG - the goal being for TestNG users to be supported and to enable use of TestNG's parallel execution support. Having a solution for TestNG is important for this project, so I'm keen to bring these in.

There's a lot of work in this, and other improvements - so thank you. 

As I read this though, I'm wondering if there might have be some more structural changes I should make to the core that would enable some of your goals to be achieved a bit more easily. I need to do some experimentation, but I can't help but feel that there could be a lighter way to run parallel tests without the need for a full grid (especially for JUnit, but perhaps for TestNG too).

Would you mind giving me a bit more time to think this over and compare it with some other options? I'm sorry to disappoint by not accepting immediately as-is; I just need to think through the best way to fit this in.

I hope this is OK

Richard


<a id='s1'>
In org.testcontainers.testng.runner.BaseTest#setUp:
    
    java.lang.AssertionError: org.openqa.selenium.WebDriverException: Error forwarding the new session Error forwarding the request Connect to 172.17.0.5:5555 [/172.17.0.5] failed: Connection refused
    Command duration or timeout: 81 milliseconds
    Build info: version: '2.52.0', revision: '4c2593cfc3689a7fcd7be52549167e5ccc93ad28', time: '2016-02-11 11:22:43'
    System info: host: 'Richards-MacBook-Pro.local', ip: '192.168.99.1', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.11.4', java.version: '1.8.0_60'
    Driver info: org.openqa.selenium.remote.RemoteWebDriver
    
        at javaslang.control.Try.getOrElseThrow(Try.java:483)
        at org.testcontainers.testng.runner.BaseTest.setUp(BaseTest.java:56)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:86)
        at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:514)
        at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:215)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:589)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:820)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1128)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
        at org.testng.TestRunner.privateRun(TestRunner.java:782)
        at org.testng.TestRunner.run(TestRunner.java:632)
        at org.testng.SuiteRunner.runTest(SuiteRunner.java:366)
        at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:361)
        at org.testng.SuiteRunner.privateRun(SuiteRunner.java:319)
        at org.testng.SuiteRunner.run(SuiteRunner.java:268)
        at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
        at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
        at org.testng.TestNG.runSuitesSequentially(TestNG.java:1244)
        at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
        at org.testng.TestNG.run(TestNG.java:1064)
        at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:74)
        at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:121)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
    Caused by: org.openqa.selenium.WebDriverException: Error forwarding the new session Error forwarding the request Connect to 172.17.0.5:5555 [/172.17.0.5] failed: Connection refused
    Command duration or timeout: 81 milliseconds
    Build info: version: '2.52.0', revision: '4c2593cfc3689a7fcd7be52549167e5ccc93ad28', time: '2016-02-11 11:22:43'
    System info: host: 'Richards-MacBook-Pro.local', ip: '192.168.99.1', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.11.4', java.version: '1.8.0_60'
    Driver info: org.openqa.selenium.remote.RemoteWebDriver
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
        at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:678)
        at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:249)
        at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:131)
        at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:158)
        at ru.stqa.selenium.factory.RemoteDriverProvider$Default.createDriver(RemoteDriverProvider.java:40)
        at ru.stqa.selenium.factory.WebDriverFactoryInternal.createRemoteDriver(WebDriverFactoryInternal.java:108)
        at ru.stqa.selenium.factory.WebDriverFactoryInternal.newDriver(WebDriverFactoryInternal.java:103)
        at ru.stqa.selenium.factory.ThreadLocalSingletonStorage.createNewDriver(ThreadLocalSingletonStorage.java:90)
        at ru.stqa.selenium.factory.ThreadLocalSingletonStorage.getDriver(ThreadLocalSingletonStorage.java:54)
        at ru.stqa.selenium.factory.WebDriverFactory.getDriver(WebDriverFactory.java:48)
        at org.testcontainers.testng.selenium.SeleniumNode.createDriverAndStartRecording(SeleniumNode.java:46)
        at org.testcontainers.testng.runner.BaseTest.lambda$setUp$12(BaseTest.java:55)
        at javaslang.control.Try.run(Try.java:56)
        at org.testcontainers.testng.runner.BaseTest.setUp(BaseTest.java:52)
        ... 30 more
    Caused by: org.openqa.grid.common.exception.GridException: Error forwarding the new session Error forwarding the request Connect to 172.17.0.5:5555 [/172.17.0.5] failed: Connection refused
        at org.openqa.grid.web.servlet.handler.RequestHandler.process(RequestHandler.java:115)
        at org.openqa.grid.web.servlet.DriverServlet.process(DriverServlet.java:83)
        at org.openqa.grid.web.servlet.DriverServlet.doPost(DriverServlet.java:67)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.seleniumhq.jetty9.servlet.ServletHolder.handle(ServletHolder.java:808)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doHandle(ServletHandler.java:587)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doHandle(SessionHandler.java:221)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
        at org.seleniumhq.jetty9.servlet.ServletHandler.doScope(ServletHandler.java:515)
        at org.seleniumhq.jetty9.server.session.SessionHandler.doScope(SessionHandler.java:185)
        at org.seleniumhq.jetty9.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
        at org.seleniumhq.jetty9.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.seleniumhq.jetty9.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.seleniumhq.jetty9.server.Server.handle(Server.java:499)
        at org.seleniumhq.jetty9.server.HttpChannel.handle(HttpChannel.java:310)
        at org.seleniumhq.jetty9.server.HttpConnection.onFillable(HttpConnection.java:257)
        at org.seleniumhq.jetty9.io.AbstractConnection$2.run(AbstractConnection.java:540)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.seleniumhq.jetty9.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.lang.Thread.run(Thread.java:745)

**Comment by sskorol** [comment]
Hi Richard,

Regarding goals:
 - To provide fully scalable Selenium Grid environment, we should not restrict users in raising only standalone images (which contain both hub and 1 particular node within the same container). This won't be a scalable solution by default. Technically there could be as many hubs / nodes as we want, and they could be located anywhere. Of course with a single hardware unit, we'll be always limited in resources to raise lots of nodes. But the main idea of grid scaling is to reduce total test execution time, and cross-browser testing. With suggested approach there was used a single hub and N configurable nodes, which could be scaled and managed by object pool. Well, instead of pooling, it'd be easier to use docker-compose scaling. But there's no such support implemented yet within this project.
 - TestNG works a bit different in comparison with JUnit from scaling perspective. In JUnit we're using forks, in TestNG - threads. That's the key difference. So I don't believe we could achieve generic approach by involving WebDriver in selenium containers' life cycle. It'd be better to let end-users to control driver on their side. And we'll just give them a test environment, where they can run tests. But such approach adds some new challanges, which I've faced with while new containers implementation. There's no easy way of how we could detect if selenium images are up and running / nodes are successfully registered on hub, etc. That's that main pitfall. To resolve this, I've added a special end-points' status check. But in such case we must know exact hub / nodes IPs / ports. And to be honest I don't know why these checks are failing on your env and CI. Perhaps I need to deeply take a look at how containers IP addresses / ports are provided on a docker level to understand the root cause of an issue. But it'd be really hard to achieve when I can't replicate it locally. This is more about docker configuration or environment itself I believe.

However, using raw selenium grid images with scaling support won't be enough for full functional testing. I'm using Grid on Windows VMs for a long time, and I know all the pitfalls and potential requirements, which users may want to see as features. Unfortunatelly, I don't have such experience with Linux and Docker yet. But I'm trying to learn all these things, as the idea of raising automation infrastructure within containers seems very promising to me. So I can help you with some Grid features + TestNG support, if you could help me to understand containers' internals deeper. :)

To summarize:
 - WebDriver is not a good candidate to be included inside containers. If we move it ouside, an issue with JUnit / TestNG support will be resolved automatically, as users will have to implement thier own runners. But it'd be fair enough. You'll provide infratrcuture and they should decide how to manage it by their own.
 - There should be added an easy way of detecting hub / nodes availability.
 - Grid is not a grid without scaling support, so it'd be nice to provide this feature at least on docker-compose level.
 - Grid containers should be linkable / mappable (shared folders) / extendable (raising additional processes within the same container via e.g. supervisord). First 2 already present in this project. And I know that last point contradicts the idea of lightweight independent containers. But from high level testing perspective it's not an occasional case, when you need to use additional tools or services to interact with remote browser (e.g. some image recognition tool - SikuliX, or some services for interacting with file system, clipboard, etc). They all should be available within a single container. Otherwise, you won't be able to get or set container-specific info / data.

I agree that current approach is too raw to be merged. So I'm ok to wait for some new low-level features to be able to improve or revise it.

Thanks,
Sergey

**Comment by bsideup** [comment]
I'm wondering what's the status of this PR :) Almost 1 year old :)

**Comment by rnorth** [comment]
Sorry I've let this languish for so long. I'm going to close it, as I think we're moving in different directions.

**Pull request closed without merging**